### ISSUE TITLE
`OneToMany::indexBy` no longer works when refering a database column

### ISSUE THREAD

**Issue by aprat84** [doctrine/doctrine2#11097]
### BC Break Report

|    Q        |   A
|------------ | ------
| BC Break    | yes
| Version     | 2.17.0

#### Summary

Indexing by database column worked, now it doesn't.
Better look at previous/current behavior.

#### Previous behavior

```php
class Language
{
    #[ORM\Id]
    #[ORM\Column(type: 'string', length: 20)]
    private ?string $id = null;
}

class Product {
    #[ORM\Id]
    #[ORM\GeneratedValue]
    #[ORM\Column(type: 'integer', options: ['unsigned' => true])]
    private ?int $id = null;

    #[ORM\OneToMany(
        mappedBy: 'product',
        targetEntity: ProductTranslation::class,
        cascade: ['persist', 'merge'],
        fetch: 'EAGER',
        indexBy: 'language_id'
    )]
    private Collection $translations;
}

class ProductTranslation {
    #[ORM\Id]
    #[ORM\ManyToOne(targetEntity: Product::class, inversedBy: 'translations')]
    #[ORM\JoinColumn(nullable: false, onDelete: 'CASCADE')]
    private Product $product;

    #[ORM\Id]
    #[ORM\ManyToOne(targetEntity: Language::class)]
    #[ORM\JoinColumn(nullable: false)]
    private Language $language;
}
```

This worked, indexing `Product::$translations` by `language_id` (`Language::$id`).

#### Current behavior

I get a warning and an error:

```txt
Warning: Undefined array key "language_id"
vendor/doctrine/orm/lib/Doctrine/ORM/Mapping/ClassMetadataInfo.php:846
```

```txt
Error: Call to a member function getValue() on null
vendor/doctrine/orm/lib/Doctrine/ORM/UnitOfWork.php:3195
```

#### How to reproduce

I've created a simple symfony project: https://github.com/aprat84/doctrine-2.17-bug

- Clone it
- Execute migrations
- Execute tests in each commit (one for 2.16.3 and one for 2.17.1)

**Comment by aprat84** [comment]
Added a repo with simple project to reproduce the bug:
https://github.com/aprat84/doctrine-2.17-bug

**Comment by igoooor** [comment]
Have you found a solution or a workaround to this? I have the same issue and don't know how to proceed

**Comment by aprat84** [comment]
This is fixed by #11380
